File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:

    def __init__(self):
        self.paths = []

    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        self.path(root,p,[])
        self.path(root,q,[])
        common_part = []
        l = min(len(self.paths[0]),len(self.paths[1]))
        i = 0
        while i < l and self.paths[0][i].val == self.paths[1][i].val:
            common_part.append(self.paths[0][i])
            i+=1

        return common_part[-1]
        

    def path(self,start,end,path):
        if start == None:
            return

        if start.val == end.val:
            self.paths.append(path + [end])
            return 

        self.path(start.left,end,path+[start])
        self.path(start.right,end,path+[start])
        return 

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if (root == nullptr) return false;  // Ran out of tree to search

        if (sameTree(root, subRoot)) return true;  // Match found at this node

        // Otherwise, check left and right subtrees
        return isSubtree(root->left, subRoot) || isSubtree(root->right, subRoot);
        
    }

    bool sameTree(TreeNode* root1, TreeNode* root2){
        if (root1 == nullptr && root2 == nullptr) {
            return true;
        }

        if( root1 == nullptr && root2 != nullptr) {
            return false;
        }

        if(root1 != nullptr && root2 == nullptr) {
            return false;
        }

        if (root1->val != root2->val) {
            return false;
        }

        return sameTree(root1->left,root2->left) && sameTree(root1->right,root2->right);
    }
};
